src/simulators/statevector/qubitvector_thrust.hpp

```diff
@@ -418,10 +418,10 @@ class QubitVectorChunkContainer
   void StoreOffsets(const std::vector<uint_t>& ptr);
 
   template <typename Function>
-  int Execute(std::vector<uint_t>& offsets,Function func,uint_t size,uint_t gid,uint_t localMask);
+  int Execute(std::vector<uint_t>& offsets,Function func,uint_t size,uint_t gid,uint_t localMask, bool omp_parallel);
```

Comment:
As the parallelism is going to affect all `Execute` calls thus we can considered it global, I'd rather add a static class state variable, and initialize it's value to one of the `thrust::omp::par` or `thrust::seq` values, depending on the expression: `omp_threshold_ < num_qubits_`, for example:
```c++
template <typename data_t>
void QubitVectorThrust<data_t>::set_num_qubits(size_t num_qubits) {
...
m_Chunks[i].enable_parallelism(omp_threshold_ < num_qubits_);
...
}
```
the implementation of `enable_parallelism`:
```c++
static decltype(thrust::seq) s_parallel_policy; // I don't know the type of thrust::seq

static void enable_parallelism(bool enable){
    s_parallel_policy = enable ? thrust::omp::par : thrust::seq;
}
...
QubitVectorChunkContainer::s_parallel_policy = thrust::seq; // Initialize static var to a safe default value
```
.. and the implementation of `Execute` and `ExecuteSum`:
```c++
...
ret = thrust::transform_reduce(s_parallel_policy, chunkIter, chunkIter + size, func,0.0,thrust::plus<double>());
...
```
This way we don't change the Execute/ExecuteSum signature, and the instances of `QubitVectorChunkContainer` will use less memory.

Reply:
This need to be part of a larger refactor i think